src/peer_manager.rs

```diff
@@ -315,11 +367,179 @@ impl PeerManager {
+    ///          (false, empty_list)        if doesn't need to carry out resource proof evaluate
+    pub fn expect_join_our_group(&mut self,
+                                 expected_name: &XorName,
+                                 our_public_id: &PublicId)
+                                 -> ExpectJoinResult {
+        if let Some((name, now, _)) = self.node_candidate {
+            if name == *expected_name ||
+               now.elapsed() < Duration::from_secs(RESOURCE_PROOF_APPROVE_TIMEOUT_SECS) {
+                return Err(RoutingTableError::AlreadyExists);
+            }
+        }
+
+        let (prefix, names) = self.routing_table.expect_join_our_group(expected_name)?;
+
+        if names.len() > self.routing_table.min_group_size() {
+            let mut rng = SeededRng::new();
+            let seed = rng.gen_iter().take(10).collect();
+            self.node_candidate = Some((*expected_name, Instant::now(), seed));
+            self.approved_candidate = None;
+        } else {
+            self.node_candidate = Some((*expected_name, Instant::now(), vec![]));
+            return Ok((false, (prefix, vec![])));
+        }
+
+        let mut public_ids = vec![];
+        for name in names {
+            if name == *our_public_id.name() {
+                public_ids.push(*our_public_id);
+            } else if let Some(peer) = self.peer_map.get_by_name(&name) {
+                public_ids.push(*peer.pub_id())
+            }
+        }
+        public_ids.sort();
+
+        Ok((true, (prefix, public_ids)))
+    }
+
+    pub fn add_as_peer_candidate(&mut self, name: XorName) {
+        self.peer_candidates.push(name)
+    }
+
+    pub fn verify_candidate(&mut self,
+                            node_name: XorName,
+                            proof: Vec<u8>,
+                            _leading_zero_bytes: u32,
+                            _difficulty: u32)
+                            -> Option<bool> {
+        if let Some((ref name, ref now, ref seed)) = self.node_candidate {
+            if *name == node_name &&
+               now.elapsed() < Duration::from_secs(RESOURCE_PROOF_EVALUATE_TIMEOUT_SECS) {
+                return Some(*seed == proof);
+            }
+        }
+        None
+    }
+
+    /// Handles node_approval vote. `validity` indicates whether rejected or approved
+    ///
+    /// Returns:
+    /// (true, Some(peer_info)) if approved peer is a node candidate
+    /// (false, None)           if approved peer is not a node candidate, or cannot find peer_info,
+    ///                         or not connected yet
+    pub fn handle_node_approval_vote(&mut self,
+                                     candidate_name: XorName,
+                                     validity: bool)
+                                     -> (bool, Option<(PublicId, PeerId)>) {
+        if let Some((name, _, _)) = self.node_candidate {
+            if name == candidate_name {
```

Comment:
as discussed, as long as the group says the joining node shall be accepted/rejected, we shall follow the decision (no matter ourself approve it or not) to maintain a group consistency. So I don't think it is necessary to checking the elapsed time here.